tsl/src/bgw_policy/policy_utils.c

```diff
@@ -79,28 +79,46 @@ subtract_integer_from_now(int64 interval, Oid time_dim_type, Oid now_func)
 		case INT2OID:
 			res = DatumGetInt16(now) - interval;
 			if (res < PG_INT16_MIN || res > PG_INT16_MAX)
-				ereport(ERROR,
-						(errcode(ERRCODE_INTERVAL_FIELD_OVERFLOW), errmsg("ts_interval overflow")));
-			return Int16GetDatum(res);
+				*overflow = true;
+			break;
 		case INT4OID:
 			res = DatumGetInt32(now) - interval;
 			if (res < PG_INT32_MIN || res > PG_INT32_MAX)
-				ereport(ERROR,
-						(errcode(ERRCODE_INTERVAL_FIELD_OVERFLOW), errmsg("ts_interval overflow")));
-			return Int32GetDatum(res);
+				*overflow = true;
+			break;
 		case INT8OID:
 		{
-			bool overflow = pg_sub_s64_overflow(DatumGetInt64(now), interval, &res);
-			if (overflow)
-			{
-				ereport(ERROR,
-						(errcode(ERRCODE_INTERVAL_FIELD_OVERFLOW), errmsg("ts_interval overflow")));
-			}
-			return Int64GetDatum(res);
+			*overflow = pg_sub_s64_overflow(DatumGetInt64(now), interval, &res);
+			break;
 		}
 		default:
 			pg_unreachable();
 	}
+	return res;
+}
+
+Datum
+subtract_integer_from_now(int64 interval, Oid time_dim_type, Oid now_func)
+{
+	AssertArg(IS_INTEGER_TYPE(time_dim_type));
+	bool overflow = false;
+	int64 res = subtract_integer_from_now_internal(interval, time_dim_type, now_func, &overflow);
+
+	if (overflow)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INTERVAL_FIELD_OVERFLOW), errmsg("ts_interval overflow")));
```

Comment:
I probably raised this before and don't recall the discussion. But do we really want to raise errors instead of just capping at boundaries? This would make the policy fail when it need not.

There are functions in `time_utils.c` now to handle saturating add and sub for time that clamp to valid time boundaries.

Reply:
Yes, this PR needs some changes after the other utility functions were added. Not yet in.

Reply:
I went back and looked at time_utils.c We have 2 versions of subtract_integer_from_now: this version throws the overflow error and used by many other policies. (Modifying other policies is outside the scope of this PR ). 

The internal version (called at line 105)  indicates the overflow and leaves it to the caller to handle as needed. This is the one used by refresh. This is needed for the refresh since the interval clamping is at the lower or upper bound  (i.e INT64_MAX or INT64_MIN ) based on whether this is the start/end interval of the refresh policy

Reply:
@erimatnor  To clarify, The refresh policy *does not throw errors*. It clamps the range.
This function is ***not*** called by refresh policy.
This is a utility function invoked by other policies. This PR does not modify the behavior of other policies.

Reply:
refresh policy calls subtract_integer_from_now_internal. This function does not throw an error.